gpudrive/datatypes/observation.py

```diff
@@ -366,3 +366,238 @@ def one_hot_encode_bev_map(self):
             self.bev_segmentation_map.long(),
             num_classes=constants.NUM_MADRONA_ENTITY_TYPES,  # From size of Madrona EntityType
         )
+
+
+@dataclass
+class TrafficLightObs:
+    """
+    A dataclass that represents traffic light information in the scenario.
+
+    This data struct contains the time series of traffic light information.
+    It contains the state (unknown, stop, caution, go), position,
+    and lane ID across the 90 recorded timesteps.
+
+    Initialized from tl_states_tensor (from Manager.trafficLightTensor()).
+    For details, see `TrafficLightState` in src/types.hpp.
+    Shape: (num_worlds, max_traffic_lights, num_timesteps, 6).
+
+    Attributes:
+        state: The state of each traffic light (0=unknown, 1=stop, 2=caution, 3=go)
+        pos_x: X-coordinate of the traffic light
+        pos_y: Y-coordinate of the traffic light
+        pos_z: Z-coordinate of the traffic light
+        time_index: Time index of the traffic light state
+        lane_id: Lane ID associated with the traffic light
+        valid_mask: Boolean mask indicating valid traffic lights
+    """
+
+    state: torch.Tensor
+    pos_x: torch.Tensor
+    pos_y: torch.Tensor
+    pos_z: torch.Tensor
+    time_index: torch.Tensor
+    lane_id: torch.Tensor
+    valid_mask: torch.Tensor
+    current_time: int = 0  # Default to the first timestep
+
+    def __init__(
+        self, tl_states_tensor: torch.Tensor, mask=None, current_time=0
+    ):
+        """Initializes the traffic light observation from a tensor."""
+        self.mask = mask
+        self.current_time = current_time
+
+        if self.mask is not None:  # Used for training
+            self.data = tl_states_tensor[self.mask]
+        else:
+            # Traffic light state (0=unknown, 1=stop, 2=caution, 3=go)
+            self.state = tl_states_tensor[:, :, :, 0].long()
+            # Position coordinates
+            self.pos_x = tl_states_tensor[:, :, :, 1]
+            self.pos_y = tl_states_tensor[:, :, :, 2]
+            self.pos_z = tl_states_tensor[:, :, :, 3]
+            # Time index and lane ID
+            self.time_index = tl_states_tensor[:, :, :, 4].long()
+            self.lane_id = tl_states_tensor[:, :, :, 5].long()
+
+            # Create a mask for valid traffic lights (lane_id != -1)
+            self.valid_mask = (self.lane_id != -1).float()
```

Comment:
what would the purpose of the valid mask be? Would we add traffic lights with invalid lane ids to the json?

Reply:
Not sure exactly, might delete the valid mask later!